src/transformers/integrations/integration_utils.py

```diff
@@ -756,6 +761,57 @@ def setup(self, args, state, model, **kwargs):
                 self._wandb.watch(model, log=_watch_model, log_freq=max(100, state.logging_steps))
             self._wandb.run._label(code="transformers_trainer")
 
+            # add number of model parameters to wandb config
+            if isinstance(
+                model,
+                (PreTrainedModel, TFPreTrainedModel, PushToHubMixin, torch.nn.Module),
+            ):
+                self._wandb.config["model/num_parameters"] = model.num_parameters()
+
+            # log the initial model and architecture to an artifact
+            with tempfile.TemporaryDirectory() as temp_dir:
+                model_name = (
+                    f"model-{self._wandb.run.id}"
+                    if (args.run_name is None or args.run_name == args.output_dir)
+                    else f"model-{self._wandb.run.name}"
+                )
+                model_artifact = self._wandb.Artifact(
+                    name=model_name,
+                    type="model",
+                    metadata={
+                        "model_config": model.config.to_dict() if hasattr(model, "config") else None,
+                        "num_parameters": model.num_parameters(),
+                        "initial_model": True,
+                    },
+                )
+                model.save_pretrained(temp_dir)
+                # add the architecture to a separate text file
+                with open(f"{temp_dir}/model_architecture.txt", "w+") as f:
+                    if isinstance(model, PreTrainedModel):
+                        print(model, file=f)
+                    elif isinstance(model, TFPreTrainedModel):
+
+                        def print_to_file(s):
+                            print(s, file=f)
+
+                        model.summary(print_fn=print_to_file)
+                    elif isinstance(model, (torch.nn.Module, PushToHubMixin)) and hasattr(model, "base_model"):
+                        print(model, file=f)
+
+                for f in Path(temp_dir).glob("*"):
```

Comment:
This seems like it could be really slow - taking all the files in the current directory and saving them out as artefacts

Reply:
This should not be an issue. This is how current artifact logging works. While the files are added to the artifact the logging happens async using the wandb service. This allows the experiment to continue while the artifact files are uploaded to W&B by the wandb service. 

Also, having all the files in a model/checkpoint allows users to make reproducible runs.